### ISSUE TITLE
GvrCardboardHelpers.SetViewerProfile

### ISSUE THREAD

**Issue by tedtheodore** [googlevr/gvr-unity-sdk#949]
This is to reopen the issue #868. It was assumed fixed in the latest version of GVR v1.150.0, but It doesn't appear to have been addressed.

I assumed the call GvrCardboardHelpers.SetViewerProfile was updated to allow for temporary viewer overrides, but either this isn't the case or some other approach was taken.

I can't find anything relating to it in the change logs.

**Comment by Daniel-Dobson** [comment]
This feature is important 👍

**Comment by tedtheodore** [comment]
I'm glad you're following too! Yeah, it's a big deal for us

**Comment by Adam-VisualVocal** [comment]
I agree, not quite fixed yet. I was able to make it work but only with some workarounds. My observations:

1. You still need to be in VR to call it. It would be much more useful if you could call it **before** entering VR.
2. We have to call `gvr_set_default_viewer_profile` ourselves instead of using the `GvrCardboardHelpers.SetViewerProfile` wrapper because it eats the bool return value which we need for step 3.
3. If `gvr_set_default_viewer_profile` succeeds (returning true), you need to call `gvr_refresh_viewer_profile` **AND** leave and reenter VR before the new default will take effect.

**Comment by Adam-VisualVocal** [comment]
Like this:
```
public static IEnumerator SetDefaultViewerProfile(string viewerProfileUri)
{
    if (XRSettings.enabled)
    {
#if (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
        // GvrCardboardHelpers.SetViewProfile does not expose the return type so we
        // call gvr_set_default_viewer_profile ourselves.

        IntPtr gvrContextPtr = GvrSettings.GetValidGvrNativePtrOrLogError();
        if (gvrContextPtr == IntPtr.Zero)
        {
            Utils.LogError("Cannot set default viewer profile! No GVR Context Ptr exists! " +
                            "You must first load AND enable a GVR device via UnityEngine.XR.XRSettings.");
            yield break;
        }

        if (gvr_set_default_viewer_profile(gvrContextPtr, viewerProfileUri))
        {
            Debug.Log("gvr_set_default_viewer_profile returned true");

            // setting the default viewer succeeded but we need to do a little dance to make it stick
            gvr_refresh_viewer_profile(gvrContextPtr);

            string deviceName = XRSettings.loadedDeviceName;

            XRSettings.enabled = false;
            XRSettings.LoadDeviceByName(string.Empty);
            yield return null;

            XRSettings.LoadDeviceByName(deviceName);
            yield return null;
            XRSettings.enabled = true;
        }
        else
        {
            Debug.Log("gvr_set_default_viewer_profile returned false");

            // the iPhone always seems to need this or it uses the wrong profile
            gvr_refresh_viewer_profile(gvrContextPtr);
        }
#else
        Debug.Log("Unavailable for non-Android and non-iOS builds");
        yield break;
#endif  // (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
    }
    else
    {
        Utils.LogError("VR must be enabled before calling SetDefaultViewerProfile.");
    }
}
```

**Comment by Ethan-VisualVocal** [comment]
👍 This needs a proper fix.  The workarounds are prone to breakage between each GVR SDK update, and cycling VR on/off for Android is a bad user experience (screen flicker).

**Comment by tedtheodore** [comment]
Thanks a lot for that! I think I will adopt this approach for the mean time and halt updates until its confirmed fixed.

**Comment by Daniel-Dobson** [comment]
Much appreciated @Adam-VisualVocal ! This is very helpful.  @Ethan-VisualVocal I fully agree yes.

**Comment by honkeat** [comment]
> 
> 
> Like this:
> 
> ```
> #if (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
>         [DllImport(GvrActivityHelper.GVR_DLL_NAME)]
>         private static extern bool gvr_set_default_viewer_profile(IntPtr gvr_context,
>             [MarshalAs(UnmanagedType.LPStr)] string viewer_profile_uri);
> 
>         [DllImport(GvrActivityHelper.GVR_DLL_NAME)]
>         private static extern void gvr_refresh_viewer_profile(IntPtr gvr_context);
> #endif
> 
> public static IEnumerator SetDefaultViewerProfile(string viewerProfileUri)
> {
>     if (XRSettings.enabled)
>     {
> #if (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
>         // GvrCardboardHelpers.SetViewProfile does not expose the return type so we
>         // call gvr_set_default_viewer_profile ourselves.
> 
>         IntPtr gvrContextPtr = GvrSettings.GetValidGvrNativePtrOrLogError();
>         if (gvrContextPtr == IntPtr.Zero)
>         {
>             Utils.LogError("Cannot set default viewer profile! No GVR Context Ptr exists! " +
>                             "You must first load AND enable a GVR device via UnityEngine.XR.XRSettings.");
>             yield break;
>         }
> 
>         if (gvr_set_default_viewer_profile(gvrContextPtr, viewerProfileUri))
>         {
>             Debug.Log("gvr_set_default_viewer_profile returned true");
> 
>             // setting the default viewer succeeded but we need to do a little dance to make it stick
>             gvr_refresh_viewer_profile(gvrContextPtr);
> 
>             string deviceName = XRSettings.loadedDeviceName;
> 
>             XRSettings.enabled = false;
>             XRSettings.LoadDeviceByName(string.Empty);
>             yield return null;
> 
>             XRSettings.LoadDeviceByName(deviceName);
>             yield return null;
>             XRSettings.enabled = true;
>         }
>         else
>         {
>             Debug.Log("gvr_set_default_viewer_profile returned false");
> 
>             // the iPhone always seems to need this or it uses the wrong profile
>             gvr_refresh_viewer_profile(gvrContextPtr);
>         }
> #else
>         Debug.Log("Unavailable for non-Android and non-iOS builds");
>         yield break;
> #endif  // (UNITY_ANDROID || UNITY_IOS) && !UNITY_EDITOR
>     }
>     else
>     {
>         Utils.LogError("VR must be enabled before calling SetDefaultViewerProfile.");
>     }
> }
> ```

Did anyone get this working? I tried the above code but it doesn't seem like the profile changed on the mobile device.
